blaze-client/src/main/scala/org/http4s/client/blaze/PoolManager.scala

```diff
@@ -1,67 +1,140 @@
+private final class PoolManager(builder: ConnectionBuilder,
+                                maxTotal: Int)
+  extends ConnectionManager {
 
   private[this] val logger = getLogger
-  private var closed = false  // All access in synchronized blocks, no need to be volatile
-  private val cs = new mutable.Queue[Connection]()
+  private var isClosed = false
+  private var allocated = 0
+  private val idleQueue = new mutable.Queue[BlazeClientStage]
+  private val waitQueue = new mutable.Queue[Callback[BlazeClientStage]]()
 
-  /** Shutdown this client, closing any open connections and freeing resources */
-  override def shutdown(): Task[Unit] = Task.delay {
-    logger.debug(s"Shutting down ${getClass.getName}.")
-    cs.synchronized {
-      if(!closed) {
-        closed = true
-        cs.foreach(_.stage.shutdown())
+  private def stats =
+    s"allocated=${allocated} idleQueue.size=${idleQueue.size} waitQueue.size=${waitQueue.size}"
+
+  private def createConnection(key: RequestKey): Unit = {
+    if (allocated < maxTotal) {
+      allocated += 1
+      logger.debug(s"Creating connection: ${stats}")
+      Task.fork(builder(key)).runAsync {
+        case \/-(stage) =>
+          logger.debug(s"Submitting fresh connection to pool: ${stats}")
+          actor ! ReturnConnection(key, stage)
+        case -\/(t) =>
+          logger.error(t)("Error establishing client connection")
+          actor ! DisposeConnection(key, None)
       }
     }
+    else {
+      logger.debug(s"Too many connections open.  Can't create a connection: ${stats}")
+    }
   }
 
-  override def recycleClient(request: Request, stage: BlazeClientStage): Unit =
-    cs.synchronized {
-      if (closed) stage.shutdown()
-      else {
-          logger.debug("Recycling connection.")
-          cs += Connection(request.uri.scheme, request.uri.authority, stage)
-
-        while (cs.size >= maxPooledConnections) {  // drop connections until the pool will fit this connection
-          logger.trace(s"Shutting down connection due to pool excess: Max: $maxPooledConnections")
-          val Connection(_, _, stage) = cs.dequeue()
-          stage.shutdown()
+  private val actor = Actor[Protocol] {
+    case RequestConnection(key, callback) =>
+      logger.debug(s"Requesting connection: ${stats}")
+      if (!isClosed) {
+        def go(): Unit = {
+          idleQueue.dequeueFirst(_.requestKey == key) match {
+            case Some(stage) if !stage.isClosed =>
+              logger.debug(s"Recycling connection: ${stats}")
+              callback(stage.right)
+            case Some(closedStage) =>
+              logger.debug(s"Evicting closed connection: ${stats}")
+              allocated -= 1
+              go()
+            case None =>
+              logger.debug(s"No connections available.  Waiting on new connection: ${stats}")
+              createConnection(key)
+              waitQueue.enqueue(callback)
```

Comment:
If we are connecting to several hosts (not my current use case), we might end up waiting because we've hit the global limit, even though we need a request to a fresh key.  We should probably close one to make room.